### PR TITLE
Ensure that 2fa is checked on reset-password

### PR DESCRIPTION
**Pull request by zeripath**

### CODE REVIEW

File: routers/user/auth.go [hunk 1]
```diff
@@ -1296,46 +1296,90 @@ func commonResetPassword(ctx *context.Context) *models.User {
...
 
 	if u == nil {
 		// Flash error has been set
 		ctx.HTML(200, tplResetPassword)
 		return
 	}
 
+	if twofa != nil {
+		useScratch := ctx.QueryBool("scratch_code")
+		if useScratch {
```
Comment (lafriks) [inline routers/user/auth.go]:
```suggestion
		if ctx.QueryBool("scratch_code") {
```

File: routers/user/auth.go [hunk 2]
```diff
@@ -1296,46 +1296,90 @@ func commonResetPassword(ctx *context.Context) *models.User {
...
+			token := ctx.Query("token")
+			ok := twofa.VerifyScratchToken(token)
+			if !ok {
```
Comment (lafriks) [inline routers/user/auth.go]:
```suggestion
			if !twofa.VerifyScratchToken(ctx.Query("token")) {
```

File: routers/user/auth.go [hunk 3]
```diff
@@ -1296,46 +1296,90 @@ func commonResetPassword(ctx *context.Context) *models.User {
...
+				ctx.Data["IsResetForm"] = true
+				ctx.Data["Err_Token"] = true
+				ctx.RenderWithErr(ctx.Tr("auth.twofa_scratch_token_incorrect"), tplResetPassword, nil)
+				return
+			}
+		} else {
+			code := ctx.Query("passcode")
+			ok, err := twofa.ValidateTOTP(code)
```
Comment (lafriks) [inline routers/user/auth.go]:
```suggestion
			ok, err := twofa.ValidateTOTP(ctx.Query("passcode"))
```

**Review by lafriks — CHANGES_REQUESTED** [review]

File: routers/user/auth.go [hunk 4]
```diff
@@ -1296,46 +1296,90 @@ func commonResetPassword(ctx *context.Context) *models.User {
...
+	}
 
 	if u == nil {
 		// Flash error has been set
 		ctx.HTML(200, tplResetPassword)
 		return
 	}
 
+	if twofa != nil {
+		useScratch := ctx.QueryBool("scratch_code")
```
Comment (zeripath) [inline routers/user/auth.go]:
```suggestion
```

File: routers/user/auth.go [hunk 5]
```diff
@@ -1296,46 +1296,90 @@ func commonResetPassword(ctx *context.Context) *models.User {
...
+		if useScratch {
+			token := ctx.Query("token")
```
Comment (zeripath) [inline routers/user/auth.go]:
```suggestion
```

File: routers/user/auth.go [hunk 6]
```diff
@@ -1296,46 +1296,90 @@ func commonResetPassword(ctx *context.Context) *models.User {
...
+			ok := twofa.VerifyScratchToken(token)
+			if !ok {
+				ctx.Data["IsResetForm"] = true
+				ctx.Data["Err_Token"] = true
+				ctx.RenderWithErr(ctx.Tr("auth.twofa_scratch_token_incorrect"), tplResetPassword, nil)
+				return
+			}
+		} else {
+			code := ctx.Query("passcode")
```
Comment (zeripath) [inline routers/user/auth.go]:
```suggestion
```

File: templates/user/auth/reset_passwd.tmpl [hunk 7]
```diff
@@ -30,10 +30,31 @@
 							</div>
 						</div>
 						{{end}}
+						{{if .has_two_factor}}
+						<h4 class="ui dividing header">
+							{{.i18n.Tr "twofa"}}
+						</h4>
+						<div class="ui warning message">{{.i18n.Tr "settings.twofa_is_enrolled" | Str2html }}</div>
```
Comment (zeripath) [inline templates/user/auth/reset_passwd.tmpl]:
```suggestion
						<div class="ui warning visible message">{{.i18n.Tr "settings.twofa_is_enrolled" | Str2html }}</div>
```

**Comment by zeripath** [comment]
So I think I need to move the scratch code and passcode check to be after the validation asif the scratch code is used you need to redirect to change it?

File: templates/user/auth/reset_passwd.tmpl [hunk 8]
```diff
@@ -30,10 +30,31 @@
 							</div>
 						</div>
 						{{end}}
+						{{if .has_two_factor}}
+						<h4 class="ui dividing header">
```
Comment (6543) [inline templates/user/auth/reset_passwd.tmpl:34]: can you indent the text?

Comment (zeripath) [reply templates/user/auth/reset_passwd.tmpl:34]: The rest of the stuff in that template does not indent.

Comment (6543) [reply templates/user/auth/reset_passwd.tmpl:34]: @zeripath why no start with this ...?

Comment (zeripath) [reply templates/user/auth/reset_passwd.tmpl:34]: we probably need to think more carefully about this indentation in templates. 

If you indent the here - the html output from the template is malindented.

**Review by 6543 — APPROVED** [review]
template format need more discusion ...
and then apply the style to all ...

but it should not block this fix

**Review by techknowlogick — APPROVED** [review]

**Review by lafriks — APPROVED** [review]

**Pull request merged**